datamol/graph.py

```diff
@@ -264,35 +267,71 @@ def reorder_mol_from_template(
             logger.warning("No match was found")
         return None
 
-    # In case we want to allow ambiguous match of hydrogens
-    if (len(matches) > 1) and allow_ambiguous_hs_only and (not allow_ambiguous_match):
-        first_keys = list(matches[0].keys())
-        all_hs_mismatch = True
-        for this_match in matches:
-            this_keys = list(this_match.keys())
-            keys_mismatch = [
-                ii for ii in range(len(this_keys)) if (first_keys[ii] != this_keys[ii])
+    if len(matches) > 1:
+        # In case we want to allow ambiguous match of hydrogens
+        if allow_ambiguous_match == "hs-only":
+            first_keys = list(matches[0].keys())
+            all_hs_mismatch = True
+            for this_match in matches:
+                this_keys = list(this_match.keys())
+                keys_mismatch = [
+                    ii for ii in range(len(this_keys)) if (first_keys[ii] != this_keys[ii])
+                ]
+                atoms_mismatch = [mol.GetAtomWithIdx(key).GetAtomicNum() for key in keys_mismatch]
+                all_hs = all([atom == 1 for atom in atoms_mismatch])
+                if not all_hs:
+                    all_hs_mismatch = False
+                    break
+            if all_hs_mismatch:
+                matches = matches[0:1]
+            else:
+                if verbose:
+                    logger.warning(
+                        f"{len(matches)} matches were found, ordering is ambiguous, even when ignoring hydrogens"
+                    )
+                return None
+
+        # Compute the number of atoms and bonds mismatch, and select the one with the least mismatch
+        if (allow_ambiguous_match in ["best", "best-first"]) and not (
+            enforce_atomic_num and enforce_bond_type
+        ):
+            num_mismatches = []
+            for this_match in matches:
+                num_atoms_mismatch = 0
+
+                # Get the number of atomic mismatch
+                for key, val in this_match.items():
+                    atom1 = mol.GetAtomWithIdx(val).GetAtomicNum()
+                    atom2 = mol_template.GetAtomWithIdx(key).GetAtomicNum()
+                    num_atoms_mismatch += atom1 != atom2
+
+                # Get the number of bond mismatch
+                num_bonds_type_mismatch, num_bonds_stereo_mismatch = 0, 0
+                for bond1 in mol_template.GetBonds():
+                    begin_idx, end_idx = bond1.GetBeginAtomIdx(), bond1.GetEndAtomIdx()
+                    bond2 = mol.GetBondBetweenAtoms(this_match[begin_idx], this_match[end_idx])
+                    num_bonds_type_mismatch += bond1.GetBondType() != bond2.GetBondType()
+                    num_bonds_stereo_mismatch += bond1.GetStereo() != bond2.GetStereo()
+
+                num_mismatches.append(
+                    num_atoms_mismatch + num_bonds_type_mismatch + num_bonds_stereo_mismatch
```

Comment:
what do you do when the two molecules have the exact same number of mismatch, but one has mismatch around atom and the other around stereo. 

I would suggest maybe a priority (random example: atom preservation >= bond type > bond stereo)  then the sort would be based on the priority to find the 'best match'.

Reply:
I changed the mismatch algorithm. I now give

- 1 point for atomic num mismatch
- 0.25 point for Formal charge mismatch
- 0.25 points for bond mismatch
- 0.05 points for stereo mismatch or bond-dir mismatch

Also, don't forget that if `enforce_atomic_num` is used, then no mismatch on the atomic number is allowed.

Does this work for you? Or do you want to change the points? Or always enforce atomic num as a priority?